File: javascript_version/dynamic_programming/overlap_subsequence/overlap_subsequence.js
// overlap subsequence
// Write a function, overlapSubsequence, that takes in two strings as arguments.
//  The function should return the length of the longest overlapping subsequence.

// A subsequence of a string can be created by deleting any characters of the string,
//  while maintaining the relative order of characters.

const overlapSubsequence = (str1, str2) => {
    // todo
    return _overlapSubsequence(str1, str2, 0, str1.length - 1, 0, str2.length - 1, {});
};


const _overlapSubsequence = (str1, str2, i = 0, j = str1.length - 1, k = 0, l = str2.length - 1, hash = {}) => {
    let key = `${i},${k}`;
    if (key in hash) { return hash[key]; }
    if (i > str1.length - 1 || k > str2.length - 1) { return 0; }
    if (str1[i] === str2[k]) {
        return 1 + _overlapSubsequence(str1, str2, i + 1, j, k + 1, l, hash);
    }
    hash[key] = Math.max(
        _overlapSubsequence(str1, str2, i + 1, j, k, l, hash),
        _overlapSubsequence(str1, str2, i, j, k + 1, l, hash));
    return hash[key];
}



console.log(overlapSubsequence("dogs", "daogt")); // -> 3
console.log(overlapSubsequence("xcyats", "criaotsi")); // -> 4
console.log(overlapSubsequence("xfeqortsver", "feeeuavoeqr")); // -> 5
console.log(overlapSubsequence("kinfolklivemustache", "bespokekinfolksnackwave")); // -> 11
console.log(overlapSubsequence(
    "mumblecorebeardleggingsauthenticunicorn",
    "succulentspughumblemeditationlocavore"
)); // -> 15

File: cpp_version/dynamic_programming/overlap_subsequence/overlap_subsequence.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits.h>
#include <math.h>
#include <queue>
#include <stack>
#include <string>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <vector>

// overlap subsequence
// Write a function, overlapSubsequence, that takes in two strings as arguments.
//  The function should return the length of the longest overlapping subsequence.

// A subsequence of a string can be created by deleting any characters of the string,
//  while maintaining the relative order of characters.

int overlapSubsequence(std::string, std::string);
int _overlapSubsequence(std::string, std::string, int, int, std::unordered_map<std::string, int> &);

int overlapSubsequence(std::string s1, std::string s2) {
    std::unordered_map<std::string, int> indexs;
    return _overlapSubsequence(s1, s2, 0, 0, indexs);
}

int _overlapSubsequence(std::string s1, std::string s2, int i, int j, std::unordered_map<std::string, int> &indexs) {
    // todo
    
        return 0;

}

int main() {

    std::cout << overlapSubsequence("dogs", "daogt") << std::endl;                                  // -> 3
    std::cout << overlapSubsequence("xcyats", "criaotsi") << std::endl;                             // -> 4
    std::cout << overlapSubsequence("xfeqortsver", "feeeuavoeqr") << std::endl;                     // -> 5
    std::cout << overlapSubsequence("kinfolklivemustache", "bespokekinfolksnackwave") << std::endl; // -> 11
    std::cout << overlapSubsequence(
                     "mumblecorebeardleggingsauthenticunicorn",
                     "succulentspughumblemeditationlocavore")
              << std::endl; // -> 15
    return 0;
}